### ISSUE TITLE
Headers management and autodoc in restx endpoints

### ISSUE THREAD

**Issue by fcamblor** [restx/restx#101]
#### Background

ATM, the simplest way to read headers from request is to inject the `RestxRequest` in our endpoint :

```
@GET("/foo")
public void foo(@Param(value="request", kind=Kind.CONTEXT) RestxRequest request) {
    Optional<String> originHeader = request.getHeader("Origin");
}
```
#### Drawbacks
- In terms of method parameter, using a `RestxRequest` is tainting the class with `Restx`-specific usage
- It is never obvious to use the `RestxRequest` trick to retrieve something
- Using `RestxRequest` means documentation will not be exhaustive regarding parameters used inside the endpoint implementation
#### Proposition
- Introduce a new context `Param.Kind` : `REQUEST_HEADER`
- It could be represented either with an `Optional<String>` or a `@NotNull String` (depending if the header should be considered as mandatory or not)
- The nice thing : we could use this special parameter kind as a documentation entry in the api-doc for  expected header name as an `in`-parameter. The form could use a dedicated parameter field for the header, then send the field value as a request header.
#### Usage

```
@GET("/foo")
public void foo(@Param(value="Origin", kind=Kind.REQUEST_HEADER) Optional<String> originHeader) {
}
```
#### Background

ATM, the only ways to write header response is :
- Either to subclass a `StdRoute` (or more generically, implement `RestxHandler`) and override the `handle(RestxRequestMatch,RestxRequest,RestxResponse,RestxContext)` method :

```
@Component
public class MyCustomRoute extends StdRoute {
    @Override
    public void handle(RestxRequestMatch match, RestxRequest req, RestxResponse resp, RestxContext ctx) throws IOException {
        MyPOJO pojo = createNewPOJO();
        resp.setHeader("Expires", DateTime.now().plusDays(2).toString());
        resp.setHeader("Location", "/foo/"+pojo.getId());
    }
}
```
- Or implement a `RestxFilter` in order to implement a more cross cutting concern like handling `ETags` or `Expires` headers. For this particular problematic (cross cutting concern), the `RestxFilter` should remain the best implementation choice.
#### Drawbacks
- Subclassing `StdRoute` means no api-docs support at all "out of the box". 
  To benefit from api-doc, we should subclass `RestxRouter` which is far more complex/verbose to describe.
- Even the `StdRoute` subclassing is far more verbose than writing a standard restx endpoint.
#### Proposition
- Introduce a new context `Param.Kind` : `RESPONSE_HEADER`
- It should be represented with a `Supplier`-like subclass, which would look like this (we might find a better name than `InstanceSupplier` here) :

```
public class InstanceSupplier<T implements Serializable> implements Supplier<Optional<T>>, Serializable {
    private Optional<T> instance = Optional.absent();
    public void set(@NotNull T instance){ this.instance = Optional.of(instance); }
    @Override
    public Optional<T> get(){ return instance; }
}
```
- The good thing : we could use this special parameter kind as a documentation entry in the api-doc for potential header name provided as an `out` parameter
#### Usage

```
@GET("/foo")
public void foo(@Param(value="Expires",kind=Kind.RESPONSE_HEADER) InstanceSupplier<String> expiresRespHeader, @Param(value="Location",kind=Kind.RESPONSE_HEADER) InstanceSupplier<String> locationRespHeader){
        expiresRespHeader.set(DateTime.now().plusDays(2).toString());
        locationRespHeader.set("/foo/"+pojo.getId());
}
```

WDYT ?

Note that I want to avoid to pass `Map` parameters because :
- A Map will be difficult to represent as a form field in the api-docs
- A map won't be self-explanatory about the fields which will be read/written from/to this map

**Comment by fedotxxl** [comment]
Hello. What's the status of this ticket? I think set response header is very basic functionality. I really don't understand why I have to extend `StdRoute` to set response header? Such a simple problem requires so much work...

**Comment by fcamblor** [comment]
@xhanin I had to read your latest comment several times to see the way you'd like to take for the `Response header` implementation in RestX.

Let me rephrase what I understood (and what are the implications) in order to see if we're on the same line.

First, you want to **avoid as much as possible** to "taint" `Resource` methods with an "HTTP API" allowing to manipulate response.

You would prefer to put response writing "outside" of the `Resource` method, in `Filters` for instance.

I had to think about it a lot before seeing how you would like to implement this easily on a per-endpoint manner (for instance having different `Expires` and `Location` response headers policies depending on the endpoint we're calling).

If I follow you well, you would implement 2 examples above by :
- Introducing new annotation (such as something like `@LocationHeader("/foo/{id}")` on a endpoint returning a POJO containing an `id` field
- Introduce a dedicated `RestxFilter` "understanding" this new `@LocationHeader` annotation and generating appropriate header value depending on the returned POJO instance

My concern, here, is I think we should plan to support non-RestX annotations (specific ones created by developers) for custom cases they would like to implement : I expect that we won't be able to cover every potential use cases for response headers.

That's why I would propose  to generate annotations metadata in the `OperationDescription`.
I just created [a PoC on my repository](https://github.com/fcamblor/restx/tree/response-headers-management) allowing to generate such metadata.
Following endpoint using a "Custom annotation" (named `@MyAnnotation`) :
```
    @GET("/testingAnnotations")
    @MyAnnotation(
            aByte=123,
            aShort=123,
            anInt=123,
            aLong=123,
            aFloat=123.456f,
            aDouble=123.456,
            aBool=true,
            aChar='A',

            // Complex types
            aString="AAA{\\\"'$AA",
            aClass=CoreResource.class,
            anEnum=MyEnum.A,

            // Another annotation
            // Not supported (yet)
            anAnnotation=@MyNestedAnnotation(value={ "BBB", "CCC" }),

            severalBytes={ 123, 123 },
            severalShorts={ 123, 456 },
            severalInts={ 123, 456 },
            severalLongs={ 123, 456 },
            severalFloats={ 123.456f, 321.654f},
            severalDoubles={ 123.456, 321.654 },
            severalBools={ true, false },
            severalChars={ 'A', 'B', 'C' },

            severalStrings={ "AAA{\\\"'$AA", "BBB", "CCC" },
            severalClasses={ Integer.class, String.class },
            severalEnums={ MyEnum.A, MyEnum.B },

            // Not supported (yet)
            severalAnnotations={
                @MyNestedAnnotation(value={ "BBB", "CCC" }),
                @MyNestedAnnotation(value={ "DDD", "EEE" })
            }
    )
    public String testingAnnotations() {
        return "hello blah";
    }
```
would generate following `describeOperation()` method in `ResourceRouter` generated class :
```
            @Override
            protected void describeOperation(OperationDescription operation) {
                super.describeOperation(operation);


                operation.responseClass = "string";
                operation.inEntitySchemaKey = "";
                operation.outEntitySchemaKey = "";
                operation.sourceLocation = "samplest.core.CoreResource#testingAnnotations()";
                operation.annotations = ImmutableList.<java.lang.annotation.Annotation>builder()
                    .add(new restx.annotations.GET() {
                        public Class<restx.annotations.GET> annotationType() { return restx.annotations.GET.class; }
                        public java.lang.String value() { return "/testingAnnotations"; }
                    })
                    .add(new samplest.annotations.MyAnnotation() {
                        public Class<samplest.annotations.MyAnnotation> annotationType() { return samplest.annotations.MyAnnotation.class; }
                        public byte aByte() { return 123; }
                        public short aShort() { return 123; }
                        public int anInt() { return 123; }
                        public long aLong() { return 123; }
                        public float aFloat() { return 123.456f; }
                        public double aDouble() { return 123.456; }
                        public boolean aBool() { return true; }
                        public char aChar() { return 'A'; }
                        public java.lang.String aString() { return "AAA{\\\"'$AA"; }
                        public java.lang.Class aClass() { return samplest.core.CoreResource.class; }
                        public samplest.annotations.MyEnum anEnum() { return samplest.annotations.MyEnum.A; }
                        public samplest.annotations.MyNestedAnnotation anAnnotation() { throw new java.lang.UnsupportedOperationException("Unsupported annotation field type"); }
                        public byte[] severalBytes() { return new byte[]{ (byte)0x7b, (byte)0x7b }; }
                        public short[] severalShorts() { return new short[]{ 123, 456 }; }
                        public int[] severalInts() { return new int[]{ 123, 456 }; }
                        public long[] severalLongs() { return new long[]{ 123L, 456L }; }
                        public float[] severalFloats() { return new float[]{ 123.456f, 321.654f }; }
                        public double[] severalDoubles() { return new double[]{ 123.456, 321.654 }; }
                        public boolean[] severalBools() { return new boolean[]{ true, false }; }
                        public char[] severalChars() { return new char[]{ 'A', 'B', 'C' }; }
                        public java.lang.String[] severalStrings() { return new java.lang.String[]{ "AAA{\\\"\'$AA", "BBB", "CCC" }; }
                        public java.lang.Class[] severalClasses() { return new java.lang.Class[]{ java.lang.Integer.class, java.lang.String.class }; }
                        public samplest.annotations.MyEnum[] severalEnums() { return new samplest.annotations.MyEnum[]{ samplest.annotations.MyEnum.A, samplest.annotations.MyEnum.B }; }
                        public samplest.annotations.MyNestedAnnotation[] severalAnnotations() { throw new java.lang.UnsupportedOperationException("Unsupported annotation field type"); }
                    })
                    .build();
            }
```

With that, we could think about storing such `OperationDescription` into a `@Component` which would be injectable into our `RestxFilter` in order to be able to have access to annotation metadata, **at runtime**, during filter processing.
It would typically allow developers to access custom annotations and make some implementations depending on those annotation, without having to use any reflection.

Obviously, `Expires` & `Location` response headers (and maybe other annotations such as `@SuccessStatus`) may be based on those annotation metadata.

Also, we would have to cover 2 sub-topics related to this new feature :
- Having to be able to "enrich" endpoint documentation generated (I would love to see those generated response headers into the api docs)
- Having to make a "matching API" for those special `RestxFilter` : enabling it only on some endpoints, or on endpoint having some custom annotations, or on endpoint returning particular Classes

WDYT about this ?

**Comment by xhanin** [comment]
Sorry for the lack of clarity, here is my feedback:

# current restx version

Here is how I would address the 2 use cases with current restx version:

## Expires
### at framework level
- define an `ExpiresStrategy` interface:
```
public interface ExpiresStrategy {
public <T> Optional<DateTime> getExpiration(Class<T> objectClass, T object);
}
```
- define a filter using an injected list of such strategies to set the expires header (see ETagFilter for an example)
### at application level
- define one or several `ExpiresStrategy` implementation depending on your use case
- endpoint methods return regular objects, handled by the `ExpiresStrategy` defined

## Location
This header should be used in 2 circumstances:
### 3xx redirect
To ask the client for a redirection with a `3xx` status code. For such case I would either return a specific object (eg `RedirectTo`) in the endpoint method, or throw a specific exception (eg `RedirectTo`), and let a filter handle the header based on either the exception catching or the object returned. The endpoint method is "tainted" with the http api, but not so much, and you can make sure nobody use the location header with the `3xx` code. 
### newly created resources
to provide location of a newly created object with a `201` or `202` status code. For such case in current restx implementation I would define a filter which test if the route is a POST, then return a `201` in case of success and add the location header. Here, without the annotation, I have to make the location deducible from the POST entry point and the object as a crosscutting concern + maybe a strategy like for `Expires`, which is not convenient if my URL paths are not always clean.

Eg if I POST on `/foos` I can automatically build a `Location` of `/foos/{id}` and get the id in response entity object, but if people define an API where you POST on `/foo` and need to redirect to `/bars/{id}` then you need to have something very specific which is not convenient at all in current implementation because you need to define a strategy implementation handling all cases and this is cumbersome and the code defining the location is not colocated with the code producing the result.

## Conclusion

Except for the POST with no naming conventions on your locations, I think you can properly address the location and expires use cases with current restx implementation, and avoid spreading the responsibility of actually setting the headers all over the application endpoint methods code. I think it kind of "forces" a cleaner design.

#  your proposal

I really like your proposal of allowing to define methods on endpoint methods and accessing such metadata in the runtime and the documentation. It allows to address the case of POST without a clean naming strategy with a much more convenient and easy way. And it can be used in a much broader kind of use cases, it's very flexible. So I'm a big +1 for it, very nice idea!

# should we support setting the headers directly in the endpoint methods?

I see that several people have voted for it, and even though I prefer to address these use cases with the approach I suggest with the benefit of your addition Fred, maybe we can listen to the community and let people have access to `RestxResponse` in their endpoint method parameters as they do for `RestxRequest`, with a kind of "use at your own risk" disclaimer... 


WDYT?

**Comment by fcamblor** [comment]
# Expires header

I [just published an implementation for a new `@ExpiresAfter` annotation header](https://github.com/fcamblor/restx/commits/response-headers-management), handled with a dedicated `ExpiresHeaderFilter` thanks to the annotation metadata I spoke about yesterday.

## Annotation usage :
```
    @GET("/expires")
    @ExpiresAfter("2d 4h")
    public String expireHeader() {
        return "hello";
    }
```

## Filter implementation :
```
@Component
public class ExpiresHeaderFilter extends EntityRelatedFilter {

    public ExpiresHeaderFilter() {
        super(Predicates.<StdRoute>alwaysTrue(), Predicates.<ResourceDescription>alwaysTrue(),
                new OperationDescription.Matcher().havingAnyAnnotations(ExpiresAfter.class)
        );
    }

    @Override
    protected void onEntityOutput(StdRoute stdRoute, RestxRequest req, RestxResponse resp,
                                  Optional<?> input, Optional<?> output,
                                  ResourceDescription resourceDescription, OperationDescription operationDescription) {

        if(!output.isPresent()) {
            return;
        }

        ExpiresAfter expiresAfterAnn = operationDescription.findAnnotation(ExpiresAfter.class).get();

        PeriodParser parser = new PeriodFormatterBuilder()
                .appendYears().appendSuffix("y").appendSeparatorIfFieldsAfter(" ")
                .appendMonths().appendSuffix("mo").appendSeparatorIfFieldsAfter(" ")
                .appendWeeks().appendSuffix("w").appendSeparatorIfFieldsAfter(" ")
                .appendDays().appendSuffix("d").appendSeparatorIfFieldsAfter(" ")
                .appendHours().appendSuffix("h").appendSeparatorIfFieldsAfter(" ")
                .appendMinutes().appendSuffix("m").appendSeparatorIfFieldsAfter(" ")
                .appendSeconds().appendSuffix("s")
                .toParser();

        MutablePeriod period = new MutablePeriod();
        parser.parseInto(period, expiresAfterAnn.value(), 0, Locale.US);
        String expiresHeaderValue = createRFC1123DateFormat().format(DateTime.now().plus(period).toDate());

        resp.setHeader("Expires", expiresHeaderValue);
    }

    public static DateFormat createRFC1123DateFormat() {
        SimpleDateFormat dateFormat = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z", Locale.US);
        dateFormat.setTimeZone(DateTimeZone.UTC.toTimeZone());
        return dateFormat;
    }
}
```

and ... that's all ! :-)
(still needing to add some stuff about documentation in api-docs, will think about it later)

Code looks pretty clean to me, not sure if we should keep the `ExpiresStrategy` components you were talking about earlier this morning :-)

BTW, if one wants to apply a cross cutting `Expires` header, then it will be easily achievable with a simple `RestxFilter` (we can already achieve this today).


# Location header

I agree with you with the `3xx REDIRECT` proposition.

For the "dynamic" URL used for created resource, I'm not a big fan of making the URL "deductible" of the `POST` URL, and particularly the part defining how we "identify" identifier of the new resource (always looking for a `getId()` method on returned entity being an obviously bad enforcement)

What I thought was something like this :
```
    @POST("/foos")
    @LocationHeader("/foos/{id}")
    public Foo createFoo(Foo foo) {
        return foos.insert(foo);
    }
```

Then, define a SPI allowing to locate paths to "interpolable" annotation properties (in our case, something like `restx.annotations.LocationHeader#value`)
This SPI could be loaded by annotation processor (can you confirm an annotation processor would be able to read such custom SPI defined in a restx project ... I mean, if I define such SPI in `restx-samplest`, will it be possible to take it into consideration by annotation processor while compiling `restx-samplest` ?)

If this assumption is correct, annotation processor would then be able to generate a new `StdEntityRoute` method like this :
```
@Override
public ImmutableMap<String, Object> buildAnnotationContextFor(Foo outEntity) {
    return ImmutableMap.builder()
        // Here, we would have every variables found in "interpolable" annotation properties defined in SPI, and their corresponding static call to the adequate getter
        // Note that list of interpolated fields will depend from one annotation "usage" to another (I may use @LocationHeader("/foos/{id}") in one case, and LocationHeader("/bars/{blah}/{bleh}") in another case
        .put("id", outEntity.getId())
        .build();
}
```

This "annotation context" would then be retrievable from the filters, and would be used to "interpolate" fields from annotations, having something as simple as :
```
@Component
public class LocationHeaderFilter extends EntityRelatedFilter {

    public LocationHeaderFilter() {
        super(Predicates.<StdRoute>alwaysTrue(), Predicates.<ResourceDescription>alwaysTrue(),
                new OperationDescription.Matcher().havingAnyAnnotations(Location.class)
        );
    }

    @Override
    protected void onEntityOutput(StdRoute stdRoute, RestxRequest req, RestxResponse resp,
                                  Optional<?> input, Optional<?> output,
                                  ResourceDescription resourceDescription, OperationDescription operationDescription) {

        if(!output.isPresent()) {
            return;
        }

        Location locationAnn = operationDescription.findAnnotation(Location.class).get();
        String interpolatedLocation = MoreAnnotations.interpolate(locationAnn.value(), stdRoute.buildAnnotationContextFor(output.get());
        resp.setHeader("Location", interpolatedLocation);
    }
}
```

WDYT ?

**Comment by xhanin** [comment]
# Expires header

Perfect! 

My suggestion with `ExpiresStrategy` was intended with current restx version (so without annotations) and to handle the use case with cross cutting solution: if the expires don't need to be defined per endpoint.

# Location header

About the "deductible" URL of POST, I'm not saying we should define it in restx, it's something that could be used/defined in a specific application by application developers. 

So I agree it's interesting to support the per endpoint definition, and an annotation is perfect for that. About the implementation, I have the feeling that trying to generate code for that is too complex, I think I would use reflection in this case. It's not restx "spirit", but it's already used in some third party libraries (like bean validation or Mustache), so I think I would go for it. So no need for the `buildAnnotationContext` generation, just call `interpolate` with the returned entity and let `interpolate` do the job with reflection - probably the easiest would be to convert the string in a Mustache template by doubling the curly braces and let Mustache do the interpolation.

WDYT?
